swap/log.go

```diff
@@ -0,0 +1,114 @@
+package swap
+
+import (
+	l "github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	// CallDepth is set to 1 in order to influence to reported line number of
+	// the log message with 1 skipped stack frame of calling l.Output()
+	CallDepth = 1
+	//DefaultAction is the default action filter for SwapLogs
+	DefaultAction string = "undefined"
+)
+
+// Logger wraps the ethereum logger with specific information for swap logging
+type Logger struct {
+	action string
+	logger l.Logger
+}
+
+func wrapCtx(sl Logger, ctx ...interface{}) []interface{} {
+	for _, elem := range ctx {
+		if elem == "action" && len(ctx)%2 == 0 {
+			return ctx
+		}
+	}
+	ctx = append([]interface{}{"swap_action", sl.action}, ctx...)
+	return ctx
+}
+
+// Warn is a convenient alias for log.Warn with a defined action context
+func (sl Logger) Warn(msg string, ctx ...interface{}) {
+	ctx = wrapCtx(sl, ctx...)
+	sl.logger.Warn(msg, ctx...)
+}
+
+// Error is a convenient alias for log.Error with a defined action context
+func (sl Logger) Error(msg string, ctx ...interface{}) {
+	ctx = wrapCtx(sl, ctx...)
+	sl.logger.Error(msg, ctx...)
+}
+
+//Crit is a convenient alias for log.Crit with a defined action context
+func (sl Logger) Crit(msg string, ctx ...interface{}) {
+	ctx = wrapCtx(sl, ctx...)
+	sl.logger.Crit(msg, ctx...)
+}
+
+//Info is a convenient alias for log.Info with a defined action context
+func (sl Logger) Info(msg string, ctx ...interface{}) {
+	ctx = wrapCtx(sl, ctx...)
+	sl.logger.Info(msg, ctx...)
+}
+
+//Debug is a convenient alias for log.Debug with a defined action context
+func (sl Logger) Debug(msg string, ctx ...interface{}) {
+	ctx = wrapCtx(sl, ctx...)
+	sl.logger.Debug(msg, ctx...)
+}
+
+// Trace is a convenient alias for log.Trace with a defined action context
+func (sl Logger) Trace(msg string, ctx ...interface{}) {
+	ctx = wrapCtx(sl, ctx...)
+	sl.logger.Trace(msg, ctx...)
+}
+
+// SetLogAction set the current log action prefix
+func (sl *Logger) SetLogAction(action string) {
+	//Adds default action undefined
+	if action == "" {
+		sl.action = DefaultAction
+		return
+	}
+	//Todo validate it's a specific action, if not default
+	sl.action = action
+}
+
+// NewSwapLogger is an alias for log.New
+func NewSwapLogger(overlayAddr string) (swapLogger Logger) {
+	return newLogger(overlayAddr, "")
+}
+
+// NewSwapPeerLogger is an alias for log.New
+func NewSwapPeerLogger(overlayAddr string, peerID string) (swapLogger Logger) {
+	return newLogger(overlayAddr, peerID)
+}
+
+// newLogger return a new SwapLogger Instance with ctx loaded for swap
+func newLogger(overlayAddr string, peerID string) (swapLogger Logger) {
+	swapLogger = Logger{
+		action: DefaultAction,
+	}
+	ctx := addSwapCtx(swapLogger, overlayAddr, peerID)
+	swapLogger.logger = l.New(ctx...)
+	return swapLogger
+}
+
+func addSwapCtx(sl Logger, overlayAddr string, peerID string) []interface{} {
+	ctx := []interface{}{"base", overlayAddr}
+	if peerID != "" {
```

Comment:
as i said in my previous review, i'm not a fan of this style of defensive programming.

i'd rather we just called an extra line/function for the peer logger instead of passing an empty parameter.

just my 2¢.

Reply:
It looks good to me. @mortelli , what is your problem with this?

Reply:
i don't think it's good style.

why should a function allow an empty string as a peer ID? conceptually, it is not correct. 

functions should be focused on doing one thing and doing it well. checking if `peerID` is empty means that this call is correct:

`addSwapCtx(myLogger, myAddress, "")`

[here](https://github.com/ethersphere/swarm/pull/1872#discussion_r334962257) is my previous comment about this.

Reply:
here's how you can do it with 1 less function, no `if`s and no duplicated code:

```golang
// newSwapLogger returns a new swap logger with standard context
func newSwapLogger(logPath string, overlayAddr []byte) Logger {
	ctx := []interface{}{"base", overlayAddr}
	return newLogger(logPath, ctx)
}

// newPeerLogger returns a new swap logger with a context that includes peer ID
func newPeerLogger(s *Swap, peerID enode.ID) Logger {
	ctx := []interface{}{"base", s.params.OverlayAddr[:16], "peer", peerID}
	return newLogger(s.params.LogPath, ctx)
}

func newLogger(logPath string, ctx []interface{}) Logger {
	swapLogger := Logger{
		action: DefaultAction,
	}
	swapLogger.logger = log.New(ctx...)
	setLoggerHandler(logPath, swapLogger.GetLogger())
	return swapLogger
}
```

Reply:
I like your proposal @mortelli . Having one function (`newLogger`) for both the SwapLogger and the PeerLogger is indeed confusing (in fact, it confused me). Your proposal makes it clear what kind of logger we are creating and makes the if statement redundant.

Reply:
@mortelli I refactored it as it's more concise as @Eknir stated.
Thanks for the observations